### PR TITLE
Consistently warn about vector and scalar operand mixing depreciation

### LINKED ISSUE #1890: "mixing vector and scalar operands is deprecated" doesn't hold for string operands
The following:

> global g = vector(3, 5, 7);
> print g == 7;

produces the warning `mixing vector and scalar operands is deprecated`, and prints `[F, F, T]`.  However, the following:

> global g = vector("a", "b", "c");
> print g == "b";

doesn't produce any warning (and prints `[F, T, F]`).  This seems inconsistent.

### PR DESCRIPTION
**Pull request by ezekiiel**
Resolves #1890

Previously we would only check for this during type promotion which would happen for numeric types but not for strings leading to inconsistent warning messages.

### CODE REVIEW

File: src/Expr.cc [hunk 1]
```diff
@@ -23,6 +23,8 @@
 #include "zeek/module_util.h"
 #include "zeek/script_opt/ExprOptInfo.h"
 
+#include "Type.h"
```
Comment (rsmmr) [inline src/Expr.cc:26]:
```suggestion
#include "zeek/Type.h"
```

(for consistency)

File: src/Expr.cc [hunk 2]
```diff
@@ -1249,6 +1248,16 @@ void BinaryExpr::PromoteForInterval(ExprPtr& op)
 		op = make_intrusive<ArithCoerceExpr>(op, TYPE_DOUBLE);
 	}
 
+bool BinaryExpr::IsScalarVecOp()
```
Comment (rsmmr) [inline src/Expr.cc:1251]: How about renaming to `IsScalableAggregateOp()` because the is now broader than just vectors?

File: src/Expr.cc [hunk 3]
```diff
@@ -1520,6 +1529,14 @@ AddExpr::AddExpr(ExprPtr arg_op1, ExprPtr arg_op2)
...
 		ExprError("requires arithmetic operands");
 
+	// Warn about scalar vector operations after we error about
+	// requiring arithmetic operands.
+	if ( ! IsError() && IsScalarVecOp() )
+		{
+		reporter->Warning("mixing vector and scalar operands is depreciated (%s) (%s)",
+		                  type_name(op1->GetType()->Tag()), type_name(op2->GetType()->Tag()));
+		}
+
```
Comment (rsmmr) [inline src/Expr.cc:1539]: To compress the code here and below a bit, I would suggest factoring this out into a function `CheckScalarVecOp` (or `CheckScalarAggOp`), potentially just combining it with the new `IsScalarVecOp()`.

**Review by rsmmr — COMMENTED** [review]
Looks great, just a couple suggestions on naming/style.

File: src/Expr.cc (hunk 3 above)
Comment (vpax) [reply src/Expr.cc:1539]: (Also a typo, `depreciated` -> `deprecated`)

Comment (ezekiiel) [reply src/Expr.cc]: Thanks for the feedback. Ought to be resolved in the `fixup!` commit. I'll rebase / squash if everything looks good.

**Review by rsmmr — COMMENTED** [review]
Yep, looks good to me. Just looks like there's `.stderr` baseline committed that's not needed.

**Comment by ezekiiel** [comment]
`AddTo` and friends are also subclasses of `BinaryExpr` so without adding an exception to those expressions I don't think so. This was what I tried at first but there were a couple of other ones where it caused trouble as well iirc.

**Comment by vpax** [comment]
> `AddTo` and friends are also subclasses of `BinaryExpr` so without adding an exception to those expressions I don't think so. This was what I tried at first but there were a couple of other ones where it caused trouble as well iirc.

Perhaps not worth it, but one approach would be to add a virtual predicate `bool AllowsMixedScalarVecOps()` that by default returns false but for those few returns true.  That would get to the heart of the distinction, rather than having it implicit via some replicated code

**Comment by ezekiiel** [comment]
> one approach would be to add a virtual predicate bool AllowsMixedScalarVecOps() that by default returns false but for those few returns true

Good idea. Happy to do that :)

**Comment by ezekiiel** [comment]
> Can these checks just be done by the UnaryExpr/BinaryExpr base classes?

Have actually learned something about inheritance trying to do this and I think the answer is effectively: not without more repeated code than we have here.

In an ideal world we would like to have some method on each child class which indicates if it supports scalar aggregate operations. Then, in the base class constructor we would check that attribute and raise a warning if needed. The trouble is that we can't actually [access that method](https://isocpp.org/wiki/faq/strange-inheritance#calling-virtuals-from-ctors) from the base class constructor because C++ constructs the base class first and thus the subclasses methods aren't callable at the time of the base class construction.

The suggested resolution to this in that isocpp article linked above is effectively to move to a builder pattern. That's probably not terribly realistic here so my vote is to just go with what we have here :)

**Comment by vpax** [comment]
> The trouble is that we can't actually [access that method](https://isocpp.org/wiki/faq/strange-inheritance#calling-virtuals-from-ctors) from the base class constructor

D'oh - yeah, that's bitten me too many times :-(.  Well, one other approach would be for the base class to look at the tag (since that _will_ be set already) and make its decision on that.  For my own aesthetics, that's less distasteful than duplicating the calls, given how many places have to include them, but YMMV.

**Pull request merged**